add encodeSignedUserTransaction and associated test case

wk3368 opened the pull request:
1. add encodeSignedUserTransaction in utils/tx.ts, 
    and associated test case: "encoding SignedUserTransaction hex" in encoding/index.spec.ts

2. add  another test case: "encoding TransactionPayload->ScriptFunction hex"  in encoding/index.spec.ts

wk3368 committed:
format code

wk3368 committed:
disable some eslint rules

wk3368 committed:
add description of how to fix eslint errors in README

wk3368 committed:
add the lib/runtime folder in .eslintignore, instead of add the disable rules in the specific ts files

wk3368 committed:
add test case: SignedUserTransaction

wk3368 committed:
add test case: generate TransactionPayload->ScriptFunction hex

wk3368 committed:
add a test case: "decoding SignedUserTransaction hex"

wk3368 committed:
format code by  VS Code

wk3368 committed:
maintain test cases of generate/decode TransactionPayload->ScriptFunction hex  in one file

wk3368 committed:
generate args of ScriptFunction is ok

wk3368 committed:
Step 2: generate RawUserTransaction
Step 3: generate signature of RawUserTransaction

wk3368 committed:
Step 4: generate authenticator
Step 5: generate SignedUserTransaction
test  case pass

wk3368 committed:
input sequence_number and expiration_timestamp_secs as parameters

nanne007 commented on src/utils/tx.ts:
```diff
@@ -49,3 +54,88 @@ export function encodePackage(
+  const transactionAuthenticatorVariantEd25519 = new starcoin_types.TransactionAuthenticatorVariantEd25519(public_key, signature)
+
+  // Step 5: generate SignedUserTransaction
+  const signedUserTransaction = new starcoin_types.SignedUserTransaction(rawUserTransaction, transactionAuthenticatorVariantEd25519)
```
整个方法逻辑太多了。建议拆分：
1. RawUserTransaction 的构建。
2. RawUserTransaction 的签名。
3. 签名 + RawUserTransaction => SignedUserTransaction.

wk3368 replied:
ok

wk3368 replied:
resloved in https://github.com/starcoinorg/starcoin.js/pull/10/commits/f028af8963691fde16720b508ffd9c248abbe129

nanne007 commented on src/utils/tx.ts:
```diff
@@ -49,3 +54,88 @@ export function encodePackage(
+
+  // Step 3: generate signature of RawUserTransaction
+  const HASH_PRE_FIX = 'STARCOIN::'
+  const typeName = 'RawUserTransaction'
```
可以利用 crypto_hash 目录下面的代码。

wk3368 replied:
这个地方的逻辑还稍微有一点不一样

是 sha3_256(STARCOIN:: RawUserTransaction).bytes()+ RrawUserTransactionBytes 
而不是 sha3_256(STARCOIN:: RawUserTransaction  + RrawUserTransaction).bytes()

所以我稍微改了一下crypto_hash， 加了个get_salt的方法，你看一下:       https://github.com/starcoinorg/starcoin.js/pull/10/commits/cbf71e6ea91ca33bc83215a13241fad98403bf8f

wk3368 committed:
use crypto_hash to generate hashSeed

wk3368 committed:
delete console.log

wk3368 committed:
split encodeSignedUserTransaction into 3 small functions

wk3368 committed:
1. add getNowSeconds
2. get senderSequenceNumber and expirationTimestampSecs from node, instead of hardcoding

wk3368 committed:
add "Type Architecture Diagram"

nanne007 reviewed (approved):
先这样吧。
sign 这块的代码，后面估计你还是要再抽象一下。
